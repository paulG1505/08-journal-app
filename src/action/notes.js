import Swal from "sweetalert2";
import { db } from "../firebase/firebase-config";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

export const startNewNote = () => {
    //es una promesa //funciones ya para consumir
    return async (dispatch, getState) => {
        const { uid } = getState().auth;
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }
        //documento referencial -coleccion-documento-coleccion
        const doc = await db.collection(`${uid}/journal/notes`).add(newNote);
        dispatch(activeNote(doc.id, newNote));
        dispatch(addNewNote(doc.id,newNote))
    }
}

export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note,
    },
})
export const starLoadingNotes = (uid) => {
    return async (dispatch) => {
        //espera que traiga las notas del user
        const notes = await loadNotes(uid);

        ///cambio del estado de las notas
        dispatch(setNotes(notes))
    }
}

export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: notes
})

export const startSaveNote=(note)=>{
    return async(dispatch,getState)=>{
        const { uid } = getState().auth;
        //si no viene url borramos ese campo para que no salte error
        if(!note.url){
            delete note.url
        }
        const noteToFirestore={...note};
        //elimino el id porque eso no necesito guardar
        delete noteToFirestore.id;

        await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore)

        dispatch(refreshNote(note.id,noteToFirestore))
        
        //alerta
        Swal.fire('Saved',note.title,'success')
    }
}

//estado de nuevas notas
export const addNewNote=(id,note)=>({
    type:types.notesAddNew,
    payload:{
        id,
        ...note
    }
})


//recargar entradas cuando se guarda
export const refreshNote=(id,note)=>({
    type:types.notesUpdate,
    payload:{
        id,
        note:{
            id,
            ...note
        }
    }
})

export const startUploading=(file)=>{
    return async( dispatch,getState)=>{
        const {active:note}= getState().notes//extraer nota actual
        Swal.fire({
            title:'Uploading...',
            text:'Please wait...',
            allowOutsideClick:false,
            //funcion que muestra icono cargando infinito
            onBeforeOpen:()=>{
                Swal.showLoading();
            }
        })
        const fileUrl= await fileUpload(file)
        //le guardo el url en la nota
        note.url=fileUrl
        dispatch(startSaveNote(note))
        Swal.close()//cierra la alerta cuando se acaba de subir
    }
}

export const startDeleted=(id)=>{
    return async(dispatch,getState)=>{
        const uid=getState().auth.uid;
        await db.doc(`${uid}/journal/notes/${id}`).delete();
        //eliminando del store
        dispatch(deleteNote(id))
    }
}

export const deleteNote=(id)=>({
    type:types.notesDelete,
    payload:id,
})

export const noteLogout=()=>({
    type:types.notesLogoutCleaning,
})

