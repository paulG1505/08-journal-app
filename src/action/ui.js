import { types } from "../types/types"

export const setError=(err)=>({
    type: types.uiSetError,
    payload: err,
})
export const removeError=(err)=>({
    type: types.uiRemove
})
export const uiStartLoading=(err)=>({
    type: types.uiStartLoading,
    payload:true
})

export const uiFinishLoading=(err)=>({
    type: types.uiFinishLoading,
    payload:false,
})