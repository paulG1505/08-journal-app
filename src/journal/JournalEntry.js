import React from 'react'
import moment from 'moment'
export const JournalEntry = ({ id, date, title, body, url }) => {

    const noteDate = moment(date);

    return (
        <div className="journal__entry pointer">
            {   //si el url es diferente de undefined
                url &&
                <div className="journal__entry-picture"

                    style={{
                        backgroundSize: 'cover',
                        backgroundImage: `url(${url})`
                    }}
                >
                </div>
            }
            <div className="journal__entry-body">
                <p className="journal__entry-title">
                    {title}
                </p>
                <p className="journal__entry-content">
                    {body}
                </p>
            </div>
            <div className="journal__entry-date-box">
                {/* momentjs para fechas */}
                <span>{noteDate.format('dddd')}</span>
                <h4>{noteDate.format('D')}</h4>
            </div>
        </div>
    )
}
