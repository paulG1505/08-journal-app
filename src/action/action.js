import { types } from "../types/types"
import {firebase, googleProvider} from '../firebase/firebase-config'
import { uiFinishLoading, uiStartLoading } from "./ui"
import Swal from 'sweetalert2'


//generando accion asincrona
export const startLogin=(email, password)=>{
    return(dispatch)=>{
        dispatch(uiStartLoading())
       firebase.auth().signInWithEmailAndPassword(email,password)

       .then(({user})=>{
           dispatch(login(user.uid,user.displayName))
           dispatch(uiFinishLoading())
       })
       .catch(e=>{
           dispatch(uiFinishLoading())
           //ALERTA
           Swal.fire('Error', e.message, 'error')
       })
    }
}

//login with email and password
export const startGoogleLogin=()=>{
    return(dispatch)=>{
        firebase.auth().signInWithPopup(googleProvider)
        .then(({user})=>{//desestructuracion del user de google
            dispatch(
                login(user.uid, user.displayName)
            )
        })
    }
}
//login with email register
export const startLoginWithEmail=(email,password,name)=>{
    return(dispatch)=>{
        firebase.auth().createUserWithEmailAndPassword(email,password)
        .then(async({user})=>{
            //espera obtener el usuario para modificar el nombre
            await user.updateProfile({displayName:name})
            dispatch(
                login(user.uid, user.displayName)
            )
        })
        //capturamos el error 
        .catch(e=> {
            console.log(e)
            Swal.fire('Error', e.message, 'error')
        });
    }
}

export const login=(uid,displayName)=>({
        type:types.login,
        payload:{
            uid, displayName
        }
    }
)

//logout de firebase
export const starLogout=()=>{
    return async(dispatch)=>{
        await firebase.auth().signOut();
        dispatch(logout())//restablece el estado
    }
}

export const logout=()=>({
    type:types.logout
})