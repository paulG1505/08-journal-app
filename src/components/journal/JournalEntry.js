import React from 'react'
import moment from 'moment'
import { activeNote } from '../../action/notes';
import { useDispatch } from 'react-redux';


export const JournalEntry = ({ id, date, title, body, url }) => {

    const dispatch = useDispatch();

    const noteDate = moment(date);
    const handleEntryClick = () => {
        dispatch(
            activeNote(id, { date, title, body, url }));
    }

    return (
        <div className="journal__entry pointer animate__animated animate__fadeInDownBig"
            onClick={handleEntryClick}
        >
            {   //si el url es diferente de undefined
                url &&
                <div className="journal__entry-picture"
                        
                    style={{
                        backgroundSize: 'cover',
                        backgroundImage: `url(${url})`
                    }}
                >
                </div>
            }
            <div className="journal__entry-body">
                <p className="journal__entry-title">
                    {title}
                </p>
                <p className="journal__entry-content">
                    {body}
                </p>
            </div>
            <div className="journal__entry-date-box">
                {/* momentjs para fechas */}
                <span>{noteDate.format('dddd')}</span>
                <h4>{noteDate.format('D')}</h4>
            </div>
        </div>
    )
}
