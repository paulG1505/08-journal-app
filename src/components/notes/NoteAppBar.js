import React from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { startSaveNote, startUploading } from '../../action/notes';
import moment from 'moment'

export const NoteAppBar = () => {

    
    const dispatch = useDispatch();  
    const {active} = useSelector( state => state.notes ); //extraer la nota activa
    const {date}=active
    const noteDate = moment(date);
    const handleSave=()=>{
        dispatch(startSaveNote(active))
    }
    const handlePicture=()=>{
        //seleccionamos el input oculto
        document.querySelector('#fileSelector').click()
    }

    const handleFileChange=(e)=>{
        //obtener la img
        const file= e.target.files[0];
        if (file){
            dispatch(startUploading(file));
        }
    }
    return (
        <div className="notes__appbar">
            <span> {noteDate.format('LL')}</span>

            <input
                id="fileSelector"
                type="file"
                style={{display:'none'}}
                name="file"
                onChange={handleFileChange}
            />

            <div>
                <button className="btn"
                    onClick={handlePicture}
                > Picture</button>
                <button 
                    className="btn"
                    onClick={handleSave}
                    > Save</button>
            </div>
        </div>
    )
}
