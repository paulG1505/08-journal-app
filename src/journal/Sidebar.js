import React from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { starLogout } from '../action/action';
import { startNewNote } from '../action/notes';
import { JournalEntries } from './JournalEntries'

export const Sidebar = () => {
    //useSelector para traer nuestro estado
    const { name } = useSelector(state => state.auth);


    const dispatch = useDispatch();
    const handleLogout = () => {
        dispatch(starLogout())
    }
    const handleEntry=()=>{
        dispatch(startNewNote())
    }
    return (
        <aside className="journal__sidebar">
            <div className="journal__sidebar-navbar">
                <h3 className="mt-5">
                    {/* icono de fontawsone  https://cdnjs.com/libraries/font-awesome*/}
                    <i className="far fa-moon"></i>
                    <span>  {name}</span>
                </h3>
                <button
                    onClick={handleLogout}
                    className="btn mt-5">
                    Logout
                </button>
            </div>
            <div className="journal__new-entry"
                onClick={handleEntry}
            >
                <i className="far fa-calendar-plus fa-5x"></i>
                <p className="mt-5"> New entry</p>
            </div>
            <JournalEntries />
        </aside>
    )
}
