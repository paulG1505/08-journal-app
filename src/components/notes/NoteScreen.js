import React, { useEffect, useRef } from 'react'
import {  useDispatch, useSelector } from 'react-redux'
import { activeNote, startDeleted } from '../../action/notes';
import { useForm } from '../../hooks/useForm';
import { NoteAppBar } from './NoteAppBar'

export const NoteScreen = () => {

    const dispatch = useDispatch();
    const {active:note} = useSelector(state => state.notes)//etraer estado de la nota actual
    const [formValues, handleInputChange,reset]=useForm(note);//llenar el formulario de acuerdo a la nota
    const {title,body}=formValues;//valores para los input
    //ejecuatar si el id cambia
    const activeId=useRef(note.id);//permite almacenar una variable mutable
    //para cambio de estado del useForm
    useEffect(() => {
        //si el id cambia se establece un nuevo estado
        if(note.id!==activeId.current){
            reset(note);
            activeId.current=note.id
        }
    }, [note,reset])
    //si la nota cambia esta escuchando
    useEffect(() => {
        dispatch(activeNote(formValues.id,{...formValues}))
    }, [formValues,dispatch])


    const handleDelete=()=>{
        dispatch(startDeleted(formValues.id))
    }

    return (
        <div className="notes__main-content">
            <NoteAppBar/>
            <div className="notes__content">
                <input 
                    type="text"
                    placeholder="Some awesome title"
                    className="notes__title-input"
                    name="title"
                    value={title}
                    onChange={handleInputChange}
                />
                <textarea 
                    placeholder="What happen today?"
                    className="notes__textarea"
                    value={body}
                    name='body'
                    onChange={handleInputChange}
                >
                </textarea>

              {  (note.url)&&<div className="notes__image">
                    <img
                        src={note.url}
                        alt="imagen prueba"
                    />
                </div>}

            </div>
            <button className="btn btn-danger"
                onClick={handleDelete}
            >
            Delete
            </button>
        </div>
    )
}
