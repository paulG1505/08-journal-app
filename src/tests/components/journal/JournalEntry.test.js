import React from 'react';
import configureStore from 'redux-mock-store' //ES6 modules
import thunk from 'redux-thunk' //es nuestro middleware
const { mount } = require("enzyme")
import '@testing-library/jest-dom'
import { Provider } from 'react-redux'
import { JournalEntry } from '../../../components/journal/JournalEntry';
import { activeNote } from '../../../action/notes';


const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initialState = {}
//creacion del store
let store = mockStore(initialState)
//simulacion de dispatch
store.dispatch=jest.fn()

const note={
    id:1,
    date:0,
    title:'Hola',
    body:'Mundo',
    url:'hhtps:unafoto.jpg'
}

const wrapper = mount(
    <Provider store={store}>
        
            <JournalEntry {...note}/>
        
    </Provider>
)


describe('Pruebas en journalEntry', () => {
    test('debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    })
    test('debe disparar activenote', () => {
        wrapper.find('.journal__entry').prop('onClick')();
        expect(store.dispatch).toHaveBeenCalledWith(
            activeNote(note.id,{...note})
        )
    })
    
})
