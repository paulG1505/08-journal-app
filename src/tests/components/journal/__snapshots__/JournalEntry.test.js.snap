// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en journalEntry debe mostrarse correctamente 1`] = `
<div
  className="journal__entry pointer animate__animated animate__fadeInDownBig"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(hhtps:unafoto.jpg)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Hola
    </p>
    <p
      className="journal__entry-content"
    >
      Mundo
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wednesday
    </span>
    <h4>
      31
    </h4>
  </div>
</div>
`;
